### DISCUSSION TITLE
Dynamic initial state based on machine's input

### DISCUSSION THREAD

**Discussion by JRebella** [statelyai/xstate#5105] [Ideas]
Right now the initial starting point of a machine can only be setup statically. I propose allowing setting up the initial state dynamically via a callback that takes the machine's `input` as parameter, just as you can do with `Context`:

```ts
initial: ({ input }) => {
  if (input.skipAhead) {
    return "state2"
  }
  return "state1";
},
```

I understand this might go against some principles of FSMs but the use cases are multiple and nothing super edge case

### Example use case: A multi-step wizard that can be resumed

Imagine you are building a multi-step wizard with many steps and branching paths, xstate is a great option since you can model each step as a discrete state, save entered data into context and use guards and transitions to move between the steps

Now lets say you want to periodically save data to a database so that if the wizard is quit in the middle, the user can then resume where they left off. With a dynamic initial state this becomes incredibly easy, its just a matter of gathering the persisted data, entering it into the machine via `input`, then letting the machine decide where to resume based on said `input`

```ts
initial: ({ input }) => {
  if (input.step3Data) {
    return 'step4';
  }
  if (input.step2Data) {
    return 'step3';
  }
  if (input.step1Data) {
    return 'step2';
  }

  return 'step1';
},
```

Currently (correct me if im mistaken) the only way of achieving this is by creating a dummy routing state at the start of the machine, then typing out every possible transition to every state you might want to resume at then add a guard for each transition. This adds a ton of boilerplate, is tedious to type out and maintain, and destroys the visibility of your machine's graph

Persisting the machine's state using the persistence feature is also an option but **very limiting** (saved state becomes incompatible after changes, can't be easily used on invoked machines, etc.)

**Comment by crishoj** [comment]
Perhaps you can achieve this using a factory method for the machine.

Something like...

```ts
const makeActor = (input) => {
    return setup({
        initial: determineInitialState(input),
        // ...
    }).createMachine({
        states: {
            // ...
        }
    })
}
```

**Reply by JRebella** [reply]
Overly complex tbh, should be supported natively
